src/network_asio.cpp

```diff
@@ -104,42 +118,99 @@ void connection::handle_connect(const boost::system::error_code& ec, endpoint en
 		ERR_NW << "Tried all IPs. Giving up" << std::endl;
 		throw system_error(ec);
 	} else {
-#if BOOST_VERSION >= 106600
 		LOG_NW << "Connected to " << endpoint.address() << '\n';
-#else
-		LOG_NW << "Connected to " << endpoint->endpoint().address() << '\n';
-#endif
+
+		if(endpoint.address().is_loopback()) {
+			use_tls_ = false;
+		}
 		handshake();
 	}
 }
 
 void connection::handshake()
 {
 	static const uint32_t handshake = 0;
+	static const uint32_t tls_handshake = htonl(uint32_t(1));
 
-	boost::asio::async_write(socket_, boost::asio::buffer(reinterpret_cast<const char*>(&handshake), 4),
-		std::bind(&connection::handle_write, this, std::placeholders::_1, std::placeholders::_2));
+	boost::asio::async_write(
+		*utils::get<raw_socket>(socket_),
+		boost::asio::buffer(use_tls_ ? reinterpret_cast<const char*>(&tls_handshake) : reinterpret_cast<const char*>(&handshake), 4),
+		std::bind(&connection::handle_write, this, std::placeholders::_1, std::placeholders::_2)
+	);
 
-	boost::asio::async_read(socket_, boost::asio::buffer(&handshake_response_.binary, 4),
+	boost::asio::async_read(*utils::get<raw_socket>(socket_), boost::asio::buffer(&handshake_response_.binary, 4),
 		std::bind(&connection::handle_handshake, this, std::placeholders::_1));
 }
 
 void connection::handle_handshake(const boost::system::error_code& ec)
 {
 	if(ec) {
+		if(ec == boost::asio::error::eof && use_tls_) {
+			// immediate disconnect likely means old server not supporting TLS handshake code
+			fallback_to_unencrypted();
+			return;
+		}
+
 		throw system_error(ec);
 	}
+	
+	if(use_tls_) {
+		if(handshake_response_.num == 0xFFFFFFFFU) {
```

Comment:
Would be nice to document the special handshake values we have so far and what they mean. And/or a little overview documentation on the handshake procedure. There is one [here](http://wiki.wesnoth.org/MultiplayerServerWML#The_handshake) which we can update but I think in the sources somewhere would be good too.

Reply:
The question is where? In server_base probably?

Reply:
Well, in the documentation of a function that does the handshake or the class that contains that function. So server_base sounds good.